File: apps/frontend/ios/Runner/BleManager.swift
import Foundation
import CoreBluetooth
import Flutter

class BleManager: NSObject, FlutterPlugin, CBCentralManagerDelegate, CBPeripheralManagerDelegate {
    static let serviceUUID = CBUUID(string: "0000C0DE-0000-1000-8000-00805F9B34FB")
    static let channelName = "com.example.campusconnect/ble"
    static var tids: [String] = []
    static var currentTidIndex: Int = 0
    static var tidTimer: Timer?
    static var scanTimer: Timer?
    static var centralManager: CBCentralManager?
    static var peripheralManager: CBPeripheralManager?
    static var eventSink: FlutterEventSink?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: channelName, binaryMessenger: registrar.messenger())
        let eventChannel = FlutterEventChannel(name: channelName + "/events", binaryMessenger: registrar.messenger())
        let instance = BleManager()
        registrar.addMethodCallDelegate(instance, channel: channel)
        eventChannel.setStreamHandler(instance)
    }

    // MARK: - FlutterPlugin
    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "startBleService":
            if let args = call.arguments as? [String: Any], let tids = args["tids"] as? [String] {
                BleManager.tids = tids
                BleManager.currentTidIndex = 0
                self.startAdvertising()
                self.startScanning()
                result(nil)
            } else {
                result(FlutterError(code: "INVALID_ARGS", message: nil, details: nil))
            }
        case "stopBleService":
            self.stopAdvertising()
            self.stopScanning()
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Advertising
    func startAdvertising() {
        if BleManager.peripheralManager == nil {
            BleManager.peripheralManager = CBPeripheralManager(delegate: self, queue: nil)
        }
        self.updateAdvertisingTid()
        BleManager.tidTimer?.invalidate()
        BleManager.tidTimer = Timer.scheduledTimer(withTimeInterval: 15 * 60, repeats: true) { _ in
            BleManager.currentTidIndex = (BleManager.currentTidIndex + 1) % BleManager.tids.count
            self.updateAdvertisingTid()
        }
    }

    func stopAdvertising() {
        BleManager.peripheralManager?.stopAdvertising()
        BleManager.tidTimer?.invalidate()
    }

    func updateAdvertisingTid() {
        guard !BleManager.tids.isEmpty else { return }
        let tid = BleManager.tids[BleManager.currentTidIndex]
        let advertisementData: [String: Any] = [
            CBAdvertisementDataServiceUUIDsKey: [BleManager.serviceUUID],
            CBAdvertisementDataServiceDataKey: [BleManager.serviceUUID: tid.data(using: .utf8) ?? Data()]
        ]
        BleManager.peripheralManager?.stopAdvertising()
        BleManager.peripheralManager?.startAdvertising(advertisementData)
    }

    // MARK: - Scanning
    func startScanning() {
        if BleManager.centralManager == nil {
            BleManager.centralManager = CBCentralManager(delegate: self, queue: nil)
        }
        self.scheduleScan()
    }

    func stopScanning() {
        BleManager.centralManager?.stopScan()
        BleManager.scanTimer?.invalidate()
    }

    func scheduleScan() {
        BleManager.scanTimer?.invalidate()
        BleManager.scanTimer = Timer.scheduledTimer(withTimeInterval: 5 * 60, repeats: true) { _ in
            self.performScan()
        }
        self.performScan()
    }

    func performScan() {
        BleManager.centralManager?.scanForPeripherals(withServices: [BleManager.serviceUUID], options: nil)
        DispatchQueue.main.asyncAfter(deadline: .now() + 10) {
            BleManager.centralManager?.stopScan()
            // スキャン結果をeventSinkで送信（ダミー実装）
            let dummyResult: [[String: Any]] = []
            BleManager.eventSink?(dummyResult)
        }
    }

    // MARK: - CBCentralManagerDelegate, CBPeripheralManagerDelegate
    func centralManagerDidUpdateState(_ central: CBCentralManager) {}
    func peripheralManagerDidUpdateState(_ peripheral: CBPeripheralManager) {}
}

extension BleManager: FlutterStreamHandler {
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        BleManager.eventSink = events
        return nil
    }
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        BleManager.eventSink = nil
        return nil
    }
} 

File: apps/frontend/android/app/src/main/kotlin/com/example/frontend/BleManager.kt
package com.example.frontend

import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothManager
import android.bluetooth.le.*
import android.content.Context
import android.os.Build
import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.PluginRegistry.Registrar

class BleManager(private val context: Context, private val eventSink: EventChannel.EventSink?) : MethodChannel.MethodCallHandler {
    companion object {
        const val SERVICE_UUID = "0000C0DE-0000-1000-8000-00805F9B34FB"
        var tids: List<String> = listOf()
        var currentTidIndex = 0
        var tidHandler: Handler? = null
        var scanHandler: Handler? = null
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "startBleService" -> {
                tids = (call.argument<List<String>>("tids") ?: listOf())
                currentTidIndex = 0
                startAdvertising()
                startScanning()
                result.success(null)
            }
            "stopBleService" -> {
                stopAdvertising()
                stopScanning()
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    private fun startAdvertising() {
        // TODO: 実装（BluetoothLeAdvertiserを利用）
        updateAdvertisingTid()
        tidHandler = Handler(Looper.getMainLooper())
        tidHandler?.postDelayed(object : Runnable {
            override fun run() {
                currentTidIndex = (currentTidIndex + 1) % tids.size
                updateAdvertisingTid()
                tidHandler?.postDelayed(this, 15 * 60 * 1000)
            }
        }, 15 * 60 * 1000)
    }

    private fun stopAdvertising() {
        // TODO: 実装
        tidHandler?.removeCallbacksAndMessages(null)
    }

    private fun updateAdvertisingTid() {
        // TODO: 実装
    }

    private fun startScanning() {
        scanHandler = Handler(Looper.getMainLooper())
        scanHandler?.post(object : Runnable {
            override fun run() {
                performScan()
                scanHandler?.postDelayed(this, 5 * 60 * 1000)
            }
        })
    }

    private fun stopScanning() {
        scanHandler?.removeCallbacksAndMessages(null)
    }

    private fun performScan() {
        // TODO: 実装（BluetoothLeScannerを利用）
        Handler(Looper.getMainLooper()).postDelayed({
            // スキャン結果をeventSinkで送信（ダミー実装）
            val dummyResult: List<Map<String, Any>> = listOf()
            eventSink?.success(dummyResult)
        }, 10 * 1000)
    }
} 